File: sdks/html5-javascript/lib/modules/Counter.js
/*
 *Licensed to the Apache Software Foundation (ASF) under one
 *or more contributor license agreements.  See the NOTICE file
 *distributed with this work for additional information
 *regarding copyright ownership.  The ASF licenses this file
 *to you under the Apache License, Version 2.0 (the
 *"License"); you may not use this file except in compliance
 *with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0

 *Unless required by applicable law or agreed to in writing,
 *software distributed under the License is distributed on an
 *"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *KIND, either express or implied.  See the License for the
 *specific language governing permissions and limitations
 *under the License.
 */


/*
 *  A class to model a Usergrid event.
 *
 *  @constructor
 *  @param {object} options {timestamp:0, category:'value', counters:{name : value}}
 *  @returns {callback} callback(err, event)
 */
Usergrid.Counter = function(options) {
  // var self=this;
  this._client = options.client;
  this._data = options.data || {};
  this._data.category = options.category || "UNKNOWN";
  this._data.timestamp = options.timestamp || 0;
  this._data.type = "events";
  this._data.counters = options.counters || {};
  // doCallback(callback, [false, this], this);
  //this.save(callback);
};
var COUNTER_RESOLUTIONS = [
  'all', 'minute', 'five_minutes', 'half_hour',
  'hour', 'six_day', 'day', 'week', 'month'
];
/*
 *  Inherit from Usergrid.Entity.
 *  Note: This only accounts for data on the group object itself.
 *  You need to use add and remove to manipulate group membership.
 */
Usergrid.Counter.prototype = new Usergrid.Entity();

/*
 * overrides Entity.prototype.fetch. Returns all data for counters
 * associated with the object as specified in the constructor
 *
 * @public
 * @method increment
 * @param {function} callback
 * @returns {callback} callback(err, event)
 */
Usergrid.Counter.prototype.fetch = function(callback) {
  this.getData({}, callback);
};
/*
 * increments the counter for a specific event
 *
 * options object: {name: counter_name}
 *
 * @public
 * @method increment
 * @params {object} options
 * @param {function} callback
 * @returns {callback} callback(err, event)
 */
Usergrid.Counter.prototype.increment = function(options, callback) {
  var self = this,
    name = options.name,
    value = options.value;
  if (!name) {
    return doCallback(callback, [new UsergridInvalidArgumentError("'name' for increment, decrement must be a number"), null, self], self);
  } else if (isNaN(value)) {
    return doCallback(callback, [new UsergridInvalidArgumentError("'value' for increment, decrement must be a number"), null, self], self);
  } else {
    self._data.counters[name] = (parseInt(value)) || 1;
    return self.save(callback);
  }
};
/*
 * decrements the counter for a specific event
 *
 * options object: {name: counter_name}
 *
 * @public
 * @method decrement
 * @params {object} options
 * @param {function} callback
 * @returns {callback} callback(err, event)
 */

Usergrid.Counter.prototype.decrement = function(options, callback) {
  var self = this,
    name = options.name,
    value = options.value;
  self.increment({
    name: name,
    value: -((parseInt(value)) || 1)
  }, callback);
};
/*
 * resets the counter for a specific event
 *
 * options object: {name: counter_name}
 *
 * @public
 * @method reset
 * @params {object} options
 * @param {function} callback
 * @returns {callback} callback(err, event)
 */

Usergrid.Counter.prototype.reset = function(options, callback) {
  var self = this,
    name = options.name;
  self.increment({
    name: name,
    value: 0
  }, callback);
};

/*
 * gets data for one or more counters over a given
 * time period at a specified resolution
 *
 * options object: {
 *                   counters: ['counter1', 'counter2', ...],
 *                   start: epoch timestamp or ISO date string,
 *                   end: epoch timestamp or ISO date string,
 *                   resolution: one of ('all', 'minute', 'five_minutes', 'half_hour', 'hour', 'six_day', 'day', 'week', or 'month')
 *                   }
 *
 * @public
 * @method getData
 * @params {object} options
 * @param {function} callback
 * @returns {callback} callback(err, event)
 */
Usergrid.Counter.prototype.getData = function(options, callback) {
  var start_time,
    end_time,
    start = options.start || 0,
    end = options.end || Date.now(),
    resolution = (options.resolution || 'all').toLowerCase(),
    counters = options.counters || Object.keys(this._data.counters),
    res = (resolution || 'all').toLowerCase();
  if (COUNTER_RESOLUTIONS.indexOf(res) === -1) {
    res = 'all';
  }
  start_time = getSafeTime(start);
  end_time = getSafeTime(end);
  var self = this;
  //https://api.usergrid.com/yourorgname/sandbox/counters?counter=test_counter
  var params = Object.keys(counters).map(function(counter) {
    return ["counter", encodeURIComponent(counters[counter])].join('=');
  });
  params.push('resolution=' + res);
  params.push('start_time=' + String(start_time));
  params.push('end_time=' + String(end_time));

  var endpoint = "counters?" + params.join('&');
  this._client.request({
    endpoint: endpoint
  }, function(err, data) {
    if (data.counters && data.counters.length) {
      data.counters.forEach(function(counter) {
        self._data.counters[counter.name] = counter.value || counter.values;
      });
    }
    return doCallback(callback, [err, data, self], self);
  });
};

function getSafeTime(prop) {
  var time;
  switch (typeof prop) {
    case "undefined":
      time = Date.now();
      break;
    case "number":
      time = prop;
      break;
    case "string":
      time = (isNaN(prop)) ? Date.parse(prop) : parseInt(prop);
      break;
    default:
      time = Date.parse(prop.toString());
  }
  return time;
}

File: stack/corepersistence/graph/src/main/java/org/apache/usergrid/persistence/graph/serialization/impl/shard/count/Counter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.usergrid.persistence.graph.serialization.impl.shard.count;


import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

import com.google.common.base.Preconditions;


/**
 * This class is synchronized for addition.  It is meant to be used across multiple threads
 */
public class Counter {
    /**
     * The counter to tell us how often it was invoked
     */
    private final AtomicLong invokeCounter;

    /**
     * Pointer to our "current" counter map.  We beginFlush this when time expires or we hit our count
     */
    private final ConcurrentHashMap<ShardKey, AtomicLong> counts;

    /**
     * The timestamp the concurrent map was created
     */
    private final long createTimestamp;


    /**
     * Implementation of the internal counters
     */
    public Counter() {
        this.createTimestamp = System.currentTimeMillis();
        this.invokeCounter = new AtomicLong();
        this.counts = new ConcurrentHashMap<>();
    }


    /**
     * Add the count to the key.
     */
    public void add( final ShardKey key, final long count ) {
        AtomicLong counter = counts.get( key );

        if ( counter == null ) {
            counter = new AtomicLong();
            AtomicLong existingCounter = counts.putIfAbsent( key, counter );

            if ( existingCounter != null ) {
                counter = existingCounter;
            }
        }

        counter.addAndGet( count );
        invokeCounter.incrementAndGet();
    }


    /**
     * Get the current valye from the cache
     */
    public long get( final ShardKey key ) {
        AtomicLong counter = counts.get( key );

        if ( counter == null ) {
            return 0;
        }

        return counter.get();
    }


    /**
     * Deep copy the counts from other into this counter
     * @param other
     */
    public void merge(final Counter other){

        Preconditions.checkNotNull(other, "other cannot be null");
        Preconditions.checkNotNull( other.counts, "other.counts cannot be null" );

        for(Map.Entry<ShardKey, AtomicLong> entry: other.counts.entrySet()){
            add(entry.getKey(), entry.getValue().get());
        }
    }


    /**
     * Get all entries
     * @return
     */
    public Set<Map.Entry<ShardKey, AtomicLong>> getEntries(){
        return counts.entrySet();
    }


    /**
     * Get the count of the number of times we've been incremented
     * @return
     */
    public long getInvokeCount() {
        return invokeCounter.get();
    }



    public long getCreateTimestamp() {
        return createTimestamp;
    }
}
